torch/csrc/jit/passes/autocast.cpp

```diff
@@ -81,21 +81,49 @@ c10::optional<AutocastScope> parseAutocast(Value* value) {
 void castTensorInputs(Node* node, Symbol cast_op) {
   const auto graph = node->owningGraph();
 
-  WithInsertPoint insert_point(node);
-
   std::unordered_set<Value*> casted_inputs;
   for (auto input : node->inputs()) {
     if (input->type()->kind() == TensorType::Kind) {
       casted_inputs.insert(input);
     }
   }
 
+  WithInsertPoint insert_point(node);
+
   for (auto input : casted_inputs) {
     const auto new_input = graph->insert(cast_op, {input});
     node->replaceInputWith(input, new_input);
   }
 }
 
+void updateDefaultDtype(Node* node, at::ScalarType dtype) {
```

Comment:
So here we are putting type promotion in the graph as if it's part of the original graph. Doesn't this gets baked in afterwards? e.g.
```
@torch.jit.script
def fn(x: torch.Tensor):
  return torch.softmax(x, 0)

x = torch.randn(2, 2, dtype=torch.float16)
fn(x) # fp32 output
x.to(dtype=torch.float64)
fn(x) # fp32 output
```

Reply:
This implements the same logic as eager mode. Are you concerned about the fp64->fp32 conversion?